File: src/main/java/com/adven/concordion/extensions/exam/db/commands/DBShowCommand.java
package com.adven.concordion.extensions.exam.db.commands;

import com.adven.concordion.extensions.exam.html.Html;
import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import org.concordion.api.CommandCall;
import org.concordion.api.Evaluator;
import org.concordion.api.ResultRecorder;
import org.dbunit.IDatabaseTester;
import org.dbunit.database.IDatabaseConnection;
import org.dbunit.dataset.DataSetException;
import org.dbunit.dataset.ITable;

import java.sql.SQLException;

import static com.adven.concordion.extensions.exam.html.Html.table;
import static com.google.common.base.Strings.isNullOrEmpty;
import static org.dbunit.dataset.filter.DefaultColumnFilter.includedColumnsTable;

public class DBShowCommand extends DBCommand {

    public DBShowCommand(String name, String tag, IDatabaseTester dbTester) {
        super(name, tag, dbTester);
    }

    @SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
    @Override
    public void setUp(CommandCall commandCall, Evaluator eval, ResultRecorder resultRecorder) {
        Html el = table(commandCall.getElement());
        try {
            String tableName = el.takeAwayAttr("table", eval);
            String where = el.takeAwayAttr("where", eval);
            IDatabaseConnection conn = getDbTester().getConnection();
            ITable filteredColumnsTable =
                    includedColumnsTable(isNullOrEmpty(where) ? conn.createTable(tableName)
                                                              : getFilteredTable(conn, tableName, where),
                                         parseCols(el, eval).cols()
                    );

            renderTable(el, filteredColumnsTable);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private ITable getFilteredTable(IDatabaseConnection connection, String tableName, String rowFilter) throws SQLException, DataSetException {
        return connection.createQueryTable(tableName, "SELECT * FROM " + tableName + " WHERE " + rowFilter);
    }
}

File: src/main/java/com/adven/concordion/extensions/exam/db/commands/DBShowCommand.kt
package com.adven.concordion.extensions.exam.db.commands

import com.adven.concordion.extensions.exam.core.fileExt
import com.adven.concordion.extensions.exam.core.html.html
import com.adven.concordion.extensions.exam.core.html.pre
import com.adven.concordion.extensions.exam.core.html.table
import com.adven.concordion.extensions.exam.db.DbPlugin
import com.adven.concordion.extensions.exam.db.DbTester
import com.adven.concordion.extensions.exam.db.builder.JSONWriter
import org.concordion.api.CommandCall
import org.concordion.api.Evaluator
import org.concordion.api.ResultRecorder
import org.dbunit.database.IDatabaseConnection
import org.dbunit.database.search.TablesDependencyHelper.getAllDependentTables
import org.dbunit.dataset.ITable
import org.dbunit.dataset.excel.XlsDataSet
import org.dbunit.dataset.filter.DefaultColumnFilter.includedColumnsTable
import org.dbunit.dataset.xml.FlatXmlDataSet
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Paths

class DBShowCommand(name: String, tag: String, dbTester: DbTester, valuePrinter: DbPlugin.ValuePrinter) : DBCommand(name, tag, dbTester, valuePrinter) {

    override fun setUp(cmd: CommandCall?, eval: Evaluator?, resultRecorder: ResultRecorder?) {
        val el = table(cmd.html())
        val tableName = el.takeAwayAttr("table", eval)!!
        val where = el.takeAwayAttr("where", eval)
        val saveToResources = el.takeAwayAttr("saveToResources", eval)
        val ds = el.takeAwayAttr("ds", DbTester.DEFAULT_DATASOURCE)
        val conn = dbTester.connectionFor(ds)

        renderTable(
            el,
            includedColumnsTable(
                if (where == null || where.isEmpty())
                    conn.createTable(tableName)
                else
                    getFilteredTable(conn, tableName, where),
                parseCols(el).keys.toTypedArray()
            ),
            remarks,
            valuePrinter
        )
        val dataSet = conn.createDataSet(getAllDependentTables(conn, tableName))
        val outputStream = ByteArrayOutputStream().apply {
            when(saveToResources?.fileExt() ?: "xml") {
                "json" -> JSONWriter(dataSet, this).write()
                "xls" -> XlsDataSet.write(dataSet, this)
                else -> FlatXmlDataSet.write(dataSet, this)
            }
        }
        saveIfNeeded(saveToResources, outputStream)
        el.below(pre(outputStream.toString("UTF-8")))
    }

    private fun saveIfNeeded(saveToResources: String?, outputStream: ByteArrayOutputStream) {
        saveToResources?.apply {
            FileOutputStream(
                File(Paths.get("src", "test", "resources").toFile(), this).apply {
                    parentFile.mkdirs()
                    createNewFile()
                },
                false
            ).use { outputStream.writeTo(it) }
        }
    }

    private fun getFilteredTable(connection: IDatabaseConnection, tableName: String, rowFilter: String): ITable =
        connection.createQueryTable(tableName, "SELECT * FROM $tableName WHERE $rowFilter")
}